### PR TITLE
Refactor async operations on individual package results

### PR DESCRIPTION
**Pull request by lennym**
Previously the callback for the lambda was fired before the individual results had been fully saved to s3 and emitted to SNS. This meant that results were lost from the end of the set and never returned to the client.

By refactoring to move the sending of individual results to inside the iteration over hotel ids then we can be sure that the lambda does not terminate until all results have successfully been returned to the client.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,77 +1,72 @@
...
+
+  return Promise.resolve()
+    .then(() => {
+      return normaliseParameters(event);
+    })
+    .then((params) => {
+      return searchForPackages(params);
+    })
+    .then((results) => {
+      return results.filter(r => r);
```
Comment (Kumjami) [inline index.js]: what does this line do? 

``` js
    return results.filter(r => r);
```

File: package.json [hunk 2]
```diff
@@ -34,9 +34,10 @@
   },
   "homepage": "https://github.com/numo-labs/lambda-ne-dynamic-package-provider#readme",
   "dependencies": {
-    "async": "2.0.0-rc.5",
     "aws-lambda-helper": "^2.18.0",
+    "bluebird": "^3.4.1",
```
Comment (Kumjami) [inline package.json:36]: is it worth it to include a promise library? 
can native 4.3.2 node version do the same without adding a dependency?

Comment (lennym) [inline package.json:36]: Yes. Node natively gives you `new Promise((resolve, reject) => {...})` but that is all. The library gives us things like iterators, which we use heavily.

File: index.js (hunk 1 above)
Comment (lennym) [inline index.js]: Strips the empty entries from the results.

If a hotel has no availability then it will create an empty entry in the array. We're not interested in keeping these in the results so filter them out.

**Comment by lennym** [comment]
Worth pointing out that this has been deployed as `lambda-ne-dynamic-package-provider-v2` and is listening to the search request SNS events, so if you're looking on an environment that uses the `ci` pipeline then you should already be seeing results from this provider.

**Comment by lennym** [comment]
@Kumjami This is good to go now. Sorry for the massive diff. :-(

You can largely disregard any changes in ./test/fixtures though, which amount to most of it.

**Comment by nelsonic** [comment]
Wow! You weren't kidding when you said "Big PR"... 😮 
is anything else higher priority or should I get started reviewing this now?

**Comment by lennym** [comment]
@nelsonic This is causing a real bug in prod, so should be fairly high priority.

For what it's worth, _most_ of the diff is in test fixtures, which can be fairly safely disregarded (they're just copies of the results from the real APIs).

**Comment by lennym** [comment]
Also, the new code itself isn't that big, so you may find it easier just to review the new code, rather than the changes.

File: lib/api_request.js [hunk 3]
```diff
@@ -46,22 +46,18 @@ function sort_by_price_asc (a, b) { // as a user I want to see cheapest holidays
...
-    });
-  }
+function get_hotel_info (params, callback) {
+  params.path = 'hotels';
+  var options = make_options(params);
+  AwsHelper.log.trace({ req: options, hotelId: params.hotelId }, 'Fetching hotel info');
+  http.request(options, function (err, data) {
+    /* istanbul ignore if */
+    if (err) {
+      AwsHelper.log.info({ err: err }, 'Error retrieving hotel info from API');
```
Comment (nelsonic) [inline lib/api_request.js:56]: does calling back with an error here display an error message in UI?

File: lib/api_request.js [hunk 4]
```diff
@@ -74,70 +70,30 @@ function get_hotel_info (hid, callback) {
...
+  const options = make_options(params);
+  AwsHelper.log.trace({ req: options, hotelId: params.hotelId }, 'Fetching package info');
+  http.request(options, function (err, data) {
+    if (!err && data.result && Array.isArray(data.result) && data.result.length > 0) {
+      const pkg = data.result.sort(sort_by_price_asc)[0]; // get cheapest package
+      get_hotel_info(params, function (err, hotel) {
+        /* istanbul ignore if */
+        if (err) {
+          AwsHelper.log.warn({ err: err, hotelId: params.hotelId, request: options }, 'Unable to get hotel info');
+          callback();
```
Comment (nelsonic) [inline lib/api_request.js]: should we `return callback(err);` here?

File: lib/api_request.js (hunk 3 above)
Comment (lennym) [inline lib/api_request.js:56]: No. Errors in search providers never make it back to the UI. 

We can add some error feedback to the UI, but I've been under the impression that this is not desired, and that the reasoning behind multiple providers is to be fault tolerant, and that if one provider errors we just show the results from other providers.

File: lib/api_request.js (hunk 4 above)
Comment (lennym) [inline lib/api_request.js]: Well... we don't want errors in any one particular hotel's api request pipeline to cause errors in the full result set (fault tolerance again), so we certainly need to ignore this error at some level. (_Note - [it's ignored in the current code](https://github.com/numo-labs/lambda-ne-dynamic-package-provider/blob/814cb4d38645b48eb27e3b93c5c9d11246ff3006/lib/api_request.js#L113)_)

There is an argument to be made that this module as a standalone component should throw errors as usual though, and that they should be handled at the call site. If I were building from scratch I'd probably do that, it's just as a progression from the previous iteration that it's the way it is.

tl;dr - probably yes, I'll update it.

_Edit - [updated](https://github.com/numo-labs/lambda-ne-dynamic-package-provider/pull/13/commits/18adeebda10583903b8ccdd3221fcf7546fbd96d)_

File: lib/api_request.js (hunk 3 above)
Comment (nelsonic) [inline lib/api_request.js:56]: Yeah, with multiple providers the UI should display _some_ result, 
we can re-assess if this is a requirement if we get a zero-results situation. 👍

File: lib/http_request.js [hunk 5]
```diff
@@ -9,26 +8,25 @@ var AwsHelper = require('aws-lambda-helper');
...
-    res.setEncoding('utf8');
-    var resStr = '';
-    res.on('data', function (chunk) {
-      resStr += chunk;
-    }).on('end', function () {
-      var json = null;
-      var err = null;
-      try { // avoid fatal error if ONE of the http requests has invalid JSON
-        json = JSON.parse(resStr);
+    res.pipe(bl((err, body) => {
```
Comment (nelsonic) [inline lib/http_request.js:17]: @lennym would you mind adding a comment clarifying what [`bl`](https://www.npmjs.com/package/bl) is doing here? (_thanks!_)

**Comment by nelsonic** [comment]
While I'm not a "_fan_" of using Promises, this use-case is good and the code appears to solve the issue.
merging. 👍

**Comment by lennym** [comment]
@nelsonic I really recommend becoming a Promise fan. Constructing complex async workflows is far easier, code is more readable and far less repetitive, and error handling is orders of magnitude easier.
